pkg/services/ngalert/schedule/schedule.go

```diff
@@ -407,12 +431,7 @@ func (sch *schedule) ruleEvaluationLoop(ctx context.Context) error {
 
 			// unregister and stop routines of the deleted alert rules
 			for key := range registeredDefinitions {
-				ruleInfo, ok := sch.registry.del(key)
-				if !ok {
-					sch.log.Error("unable to delete alert rule routine information because it did not exist", "uid", key.UID, "org_id", key.OrgID)
-					continue
-				}
-				ruleInfo.stop()
+				sch.DeleteAlertRule(key)
```

Comment:
Do we have a test to show that when we delete an alert rule with `DeleteAlertRule` the scheduler will still call `sch.DeleteAlertRule`? I ask because if we call `DeleteAlertRule` then won't we remove the alert rule from `sch.registry` before the scheduler has called `sch.registry.KeyMap()` such that we never call `sch.DeleteAlertRule(key)`?

Reply:
Sorry, I do not understand the use-case you describe. Can you please clarify. As for tests, there are few tests https://github.com/grafana/grafana/pull/39975/files/6af0053ca4c234154c9629c233d3246f48be0bb0#diff-bd46cfcb4ef816c767bb8b9d3637680e1c6ac38f7fd66f4d7635b06adc6d588fR866